Using CPU to create model - MacBook Pro (Retina, 13-inch, Early 2015)

I'm running the object detection task.
====================================
MacOS Version 10.14 Beta  (18A293u)

MacBook Pro  (Retina, 13-inch, Early 2015) 
Processor 2.7 GHz Intel Core i5 Memory 8 GB 1867 MHz DDR3 
Startup Disk Macintosh HD 
Graphics Intel Iris Graphics 6100 1536 MB

I tried a fresh build without running multiple make jobs:
- reinstall Xcode 10 beta on Mojave beta
- ensure command line tools are set to Xcode 10
- fresh clone of Turicreate repository

All building well until Metal comes into scope:
`Scanning dependencies of target image_deep_feature_extractor
[ 74%] Building CXX object src/toolkits/image_deep_feature_extractor/CMakeFiles/image_deep_feature_extractor.dir/mlmodel_image_feature_extractor.mm.o
In file included from /Users/xxx/Documents/turicreate/src/toolkits/image_deep_feature_extractor/mlmodel_image_feature_extractor.mm:18:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreML.framework/Headers/CoreML.h:8:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:6:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h:60:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFPropertyList.h:17:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFStream.h:16:
/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:906:91: error: 
      expected ';' after top level declarator
const CFStringRef kCFURLFileProtectionKey API_AVAILABLE(ios(9.0), watchos(2.0), tvos(9.0)) API_PROHIBITED(macos);`
multiple identical errors for missing ";" before API_PROHIBITED

--- a/src/unity/python/turicreate/toolkits/_internal_utils.py
+++ b/src/unity/python/turicreate/toolkits/_internal_utils.py
@@ -620,3 +620,27 @@ def _mac_ver():
         return tuple([int(v) for v in ver_str.split('.')])
     else:
         return ()
+
+def _print_neural_compute_device(cuda_gpus, use_mps, cuda_mem_req=None, has_mps_impl=True):
+    """
+    Print a message making it clear to the user what compute resource is used in
+    neural network training.
+    """
+    num_cuda_gpus = len(cuda_gpus)
+    if num_cuda_gpus >= 1:
+        gpu_names = ', '.join(gpu['name'] for gpu in cuda_gpus)
+
+    if use_mps:
+        from ._mps_utils import mps_device_name
+        print('Using GPU to create model ({})'.format(mps_device_name()))
+    elif num_cuda_gpus >= 1:
+        from . import _mxnet_utils
+        plural = 's' if num_cuda_gpus >= 2 else ''
+        print('Using GPU{} to create model ({})'.format(plural, gpu_names))
+        if cuda_mem_req is not None:
+            _mxnet_utils._warn_if_less_than_cuda_free_memory(cuda_mem_req, max_devices=num_cuda_gpus)
+    else:
+        import sys
+        print('Using CPU to create model')
+        if sys.platform == 'darwin' and _mac_ver() < (10, 14) and has_mps_impl:
+            print('NOTE: If available, an AMD GPU can be leveraged on macOS 10.14+ for faster model creation')

--- a/src/unity/python/turicreate/toolkits/_mxnet_utils.py
+++ b/src/unity/python/turicreate/toolkits/_mxnet_utils.py
@@ -44,20 +44,30 @@ def get_mxnet_state(model):
     return state
 
 
+def get_gpu_ids_in_use(max_devices=None):
+    from turicreate.util import _CUDA_GPUS
+    gpus = _CUDA_GPUS
+    num_gpus = get_num_gpus_in_use(max_devices=max_devices)
+    if num_gpus >= 0:
+        gpus = gpus[:num_gpus]
+    if max_devices is not None:
+        gpus = gpus[:max_devices]
+    return [gpu_info['index'] for gpu_info in gpus]
+
+
 def get_mxnet_context(max_devices=None):
-    from turicreate.util import _CUDA_GPU_IDS
+    from turicreate.util import _CUDA_GPUS
     import mxnet as _mx
     assert_valid_num_gpus()
     num_gpus = _tc_config.get_num_gpus()
-    if num_gpus == 0 or not _CUDA_GPU_IDS:
+    if num_gpus == 0 or not _CUDA_GPUS:
         return [_mx.cpu()]
     else:
-        if num_gpus == -1:
-            ctx = [_mx.gpu(i) for i in _CUDA_GPU_IDS]
-        elif num_gpus > 0:
-            ctx = [_mx.gpu(i) for i in range(num_gpus)]
+        gpu_indices = [gpu['index'] for gpu in _CUDA_GPUS]
+        if num_gpus > 0:
+            gpu_indices = gpu_indices[:num_gpus]
 
-        ctx = ctx[:max_devices]
+        ctx = [_mx.gpu(index) for index in gpu_indices]
 
         # Do a quick check if we can use the GPU
         try:
@@ -87,16 +97,86 @@ def get_num_gpus_in_use(max_devices=None):
     return len(gpu_ctx)
 
 
+def get_gpus_in_use(max_devices=None):
+    """
+    Like get_num_gpus_in_use, but returns a list of dictionaries with just
+    queried GPU information.
+    """
+    from turicreate.util import _get_cuda_gpus
+    gpu_indices = get_gpu_ids_in_use(max_devices=max_devices)
+    gpus = _get_cuda_gpus()
+    return [gpus[index] for index in gpu_indices]
+
+
 def assert_valid_num_gpus():
-    from turicreate.util import _CUDA_GPU_IDS
+    from turicreate.util import _CUDA_GPUS
     num_gpus = _tc_config.get_num_gpus()
-    if not _CUDA_GPU_IDS and _sys.platform == 'darwin':
+    if not _CUDA_GPUS and _sys.platform == 'darwin':
         # GPU acceleration requires macOS 10.14+
         if num_gpus == 1 and _mac_ver() < (10, 14):
             raise _ToolkitError('GPU acceleration requires at least macOS 10.14')
         elif num_gpus >= 2:
             raise _ToolkitError('Using more than one GPU is currently not supported on Mac')
     _numeric_param_check_range('num_gpus', num_gpus, -1, _six.MAXSIZE)
+
+
+def _warn_if_less_than_cuda_free_memory(mem_mb, max_devices=None, has_batch_size_setting=True):
+    # Note, we cannot use _CUDA_GPUS, since we need an up-to-date snapshot of
+    # the free memory
+    import numpy as np
+    from turicreate.util import _get_cuda_gpus
+    import time
+    import textwrap
+    gpus = _get_cuda_gpus()
+    num_gpus = get_num_gpus_in_use(max_devices=max_devices)
+    gpus = gpus[:num_gpus]
+    capable = all(gpu['memory_total'] > mem_mb for gpu in gpus)
+    available = all(gpu['memory_free'] > mem_mb for gpu in gpus)
+    warning = None
+    if capable and not available:
+        # Get GPU with the least amount of free memroy
+        gpu_mems = [gpu['memory_free'] for gpu in gpus]
+        gpu_idx = np.argmin(gpu_mems)
+
+        warning = '''
+        You may not have enough free GPU memory to create this model, which may
+        result in a fatal crash or lower performance. We recommend that you
+        have at least {mem_req} MiB available on each GPU, while currently you
+        have only {mem_free:.0f} MiB free (out of {mem_total:.0f} MiB) on GPU
+        #{index:d}. Please close all other processes that may be using this
+        GPU.
+        '''.format(mem_req=mem_mb,
+                   mem_free=gpus[gpu_idx]['memory_free'],
+                   mem_total=gpus[gpu_idx]['memory_total'],
+                   index=gpus[gpu_idx]['index'])
+
+    elif not capable:
+        # Get GPU with the least amount of total memroy
+        gpu_mems = [gpu['memory_total'] for gpu in gpus]
+        gpu_idx = np.argmin(gpu_mems)
+
+        warning = '''
+        You may not have a capable enough GPU to create this model with the
+        current settings, which may result in a fatal crash or lower
+        performance. We recommend that you have at least {mem_req} MiB on each
+        requested GPU, while currently GPU #{index} has only {mem_total:.0f}
+        MiB of total memory.
+        '''.format(mem_req=mem_mb,
+                   mem_total=gpus[gpu_idx]['memory_total'],
+                   index=gpus[gpu_idx]['index'])
+
+        if has_batch_size_setting:
+            warning += 'Please try lowering the batch size by passing the parameter batch_size to create().'
+
+    if warning:
+        print('WARNING:')
+        print()
+        # Re-wrap text
+        lines = textwrap.wrap(textwrap.dedent(warning).lstrip(), width=75)
+        for line in lines:
+            print('   ', line)
+        print()
+        time.sleep(3)
 
 
 def load_mxnet_model_from_state(state, data, labels=None, existing_module=None, ctx = None):

--- a/src/unity/python/turicreate/toolkits/_mxnet_utils.py
+++ b/src/unity/python/turicreate/toolkits/_mxnet_utils.py
@@ -122,8 +122,5 @@ def get_gluon_net_params_state(net_params):
 def load_net_params_from_state(net_params, state, ctx = None):
     net_params_dict = params_from_dict(state, ctx)
     for k in net_params_dict:
-        #net_params[k].initialize(ctx)
-        #import pdb; pdb.set_trace()
-        #net_params[k].set_data(net_params_dict[k])
         net_params[k]._load_init(net_params_dict[k], ctx)
     return net_params

--- a/src/unity/python/turicreate/toolkits/object_detector/object_detector.py
+++ b/src/unity/python/turicreate/toolkits/object_detector/object_detector.py
@@ -31,7 +31,6 @@ from turicreate.toolkits._main import ToolkitError as _ToolkitError
 from .. import _pre_trained_models
 from ._evaluation import average_precision as _average_precision
 from .._mps_utils import (use_mps as _use_mps,
-                          mps_device_name as _mps_device_name,
                           mps_device_memory_limit as _mps_device_memory_limit,
                           MpsGraphAPI as _MpsGraphAPI,
                           MpsGraphNetworkType as _MpsGraphNetworkType,
@@ -272,7 +271,8 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
 
     if batch_size < 1:
         batch_size = 32  # Default if not user-specified
-    num_mxnet_gpus = _mxnet_utils.get_num_gpus_in_use(max_devices=batch_size)
+    cuda_gpus = _mxnet_utils.get_gpus_in_use(max_devices=batch_size)
+    num_mxnet_gpus = len(cuda_gpus)
     use_mps = _use_mps() and num_mxnet_gpus == 0
     batch_size_each = batch_size // max(num_mxnet_gpus, 1)
     if use_mps and _mps_device_memory_limit() < 4 * 1024 * 1024 * 1024:
@@ -291,14 +291,11 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
     io_thread_buffer_size = params['io_thread_buffer_size'] if use_mps else 0
 
     if verbose:
-        if use_mps:
-            print('Using GPU to create model ({})'.format(_mps_device_name()))
-        elif num_mxnet_gpus == 1:
-            print('Using GPU to create model (CUDA)')
-        elif num_mxnet_gpus > 1:
-            print('Using {} GPUs to create model (CUDA)'.format(num_mxnet_gpus))
-        else:
-            print('Using CPU to create model')
+        # Estimate memory usage (based on experiments)
+        cuda_mem_req = 550 + batch_size_each * 85
+
+        _tkutl._print_neural_compute_device(cuda_gpus=cuda_gpus, use_mps=use_mps,
+                                            cuda_mem_req=cuda_mem_req)
 
     grid_shape = params['grid_shape']
     input_image_shape = (3,
@@ -558,6 +555,8 @@ def create(dataset, annotations=None, feature=None, model='darknet-yolo',
                 net_params[k].set_data(_mps_to_mxnet(mps_net_params[k]))
 
     else:  # Use MxNet
+        net.hybridize()
+
         options = {'learning_rate': base_lr, 'lr_scheduler': lr_scheduler,
                    'momentum': params['sgd_momentum'], 'wd': params['weight_decay'], 'rescale_grad': 1.0}
         clip_grad = params.get('clip_gradients')

--- a/src/unity/python/turicreate/toolkits/style_transfer/style_transfer.py
+++ b/src/unity/python/turicreate/toolkits/style_transfer/style_transfer.py
@@ -227,15 +227,15 @@ def create(style_dataset, content_dataset, style_feature=None,
     smoothed_loss = None
     last_time = 0
 
-    num_mxnet_gpus = _mxnet_utils.get_num_gpus_in_use(max_devices=params['batch_size'])
-    if verbose:
-        if num_mxnet_gpus == 1:
-            print('Using GPU to create model (CUDA)')
-        elif num_mxnet_gpus > 1:
-            print('Using {} GPUs to create model (CUDA)'.format(num_mxnet_gpus))
-        else:
-            print('Using CPU to create model')
+    cuda_gpus = _mxnet_utils.get_gpus_in_use(max_devices=params['batch_size'])
+    num_mxnet_gpus = len(cuda_gpus)
 
+    if verbose:
+        # Estimate memory usage (based on experiments)
+        cuda_mem_req = 260 + batch_size_each * 880 + num_styles * 1.4
+
+        _tkutl._print_neural_compute_device(cuda_gpus=cuda_gpus, use_mps=False,
+                                            cuda_mem_req=cuda_mem_req, has_mps_impl=False)
     #
     # Pre-compute gram matrices for style images
     #

--- a/src/unity/python/turicreate/util/__init__.py
+++ b/src/unity/python/turicreate/util/__init__.py
@@ -717,19 +717,41 @@ def subprocess_exe(exe, args, setup=None, teardown=None,
 # Automatic GPU detection
 def _get_cuda_gpus():
     """
-    Returns a list of 0-based integer indices of available CUDA GPUs.
+    Returns a list of dictionaries, with the following keys:
+    - index (integer, device index of the GPU)
+    - name (str, GPU name)
+    - memory_free (float, free memory in MiB)
+    - memory_total (float, total memory in MiB)
     """
     import subprocess
     try:
-        ret = subprocess.check_output(["nvidia-smi", "-L"], universal_newlines=True).split('\n')
-        return [i for i, s in enumerate(ret) if 'GPU' in s]
+        output = subprocess.check_output(['nvidia-smi',
+                                          '--query-gpu=index,gpu_name,memory.free,memory.total',
+                                          '--format=csv,noheader,nounits'],
+                                         universal_newlines=True)
     except OSError:
         return []
-_CUDA_GPU_IDS = _get_cuda_gpus()
+
+    gpus = []
+    for gpu_line in output.split('\n'):
+        if gpu_line:
+            index, gpu_name, memory_free, memory_total = gpu_line.split(', ')
+            index = int(index)
+            memory_free = float(memory_free)
+            memory_total = float(memory_total)
+            gpus.append({
+                'index': index,
+                'name': gpu_name,
+                'memory_free': memory_free,
+                'memory_total': memory_total,
+            })
+    return gpus
+
+_CUDA_GPUS = _get_cuda_gpus()
 
 
 def _num_available_cuda_gpus():
-    return len(_CUDA_GPU_IDS)
+    return len(_CUDA_GPUS)
 
 
 def _num_available_gpus():

